urlc8dbc7ce6e_dlubarov_guava_tgz-pJ8-MainJ8/src/c/TypeDef.java

```diff
@@ -0,0 +1,441 @@
+    private Set<MethodDef> allInheritedInstanceMethods() {
+        Set<MethodDef> methods = new HashSet<MethodDef>();
+        for (MethodDef m : instanceMethodDefs)
+            methods.add(m);
+        for (ParameterizedType parent : parents) {
+            TypeDef parentDef = Project.singleton.resolve(parent.rawType);
+            methods.addAll(parentDef.allInheritedInstanceMethods());
+        }
+        return methods;
+    }
+
+    private boolean hasSupertype(RawType sup) {
+        try {
+            thisType().asSupertype(sup, new CodeContext(this, null));
+            return true;
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+    }
+
+    // Find my implementation of an instance method.
+    private MethodDef getImplementation(MethodDef method) {
+        if (!hasSupertype(method.owner))
+            throw new NoSuchElementException("This should be caught...");
+
+        // Try my instance methods.
+        Set<MethodDef> options = new HashSet<MethodDef>();
+        for (MethodDef m : instanceMethodDefs)
+            if (m.canImplement(method))
+                options.add(m);
+
+        // If nothing was found yet, try my parents.
+        if (options.isEmpty())
+            for (ParameterizedType parent : parents) {
+                TypeDef parentDef = Project.singleton.resolve(parent.rawType);
+                try {
+                    options.add(parentDef.getImplementation(method));
+                } catch (NoSuchElementException e) {}
+            }
+
+        if (options.isEmpty())
+            throw new NoSuchElementException(String.format(
+                    "Type '%s' has no implementation of '%s.%s'.",
+                    desc, method.owner, method.name));
+
+        // Attempt to narrow down to 1 with canImplement checks.
+        // Example: Suppose Sequence[T] overrides Top.hashCode while Immutable inherits it,
+        // and BitSet implements Sequence[Bool] and Immutable. Then BitSet will have two
+        // implementations of hashCode, Top's and Sequence's. This code should eliminate
+        // Top.hashCode from options.
+        for (MethodDef m1 : new HashSet<MethodDef>(options))
+            for (MethodDef m2 : options)
+                // If m2 overrides m1, remove m1 from options.
+                if (m2 != m1 && m2.canImplement(m1)) {
+                    options.remove(m1);
+                    break;
+                }
+
+        if (options.size() > 1) {
+            String[] optionDescs = new String[options.size()];
+            int i = 0;
+            for (MethodDef m : options)
+                optionDescs[i++] = m.owner + " " + m.name;
+            throw new NiftyException(
+                    "Type '%s' has multiple implementations of '%s.%s': %s.",
+                    desc, method.owner, method.name, Arrays.toString(optionDescs));
+        }
+
+        return options.iterator().next();
+    }
+
+    public d.TypeDef compile() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 314-321 into a separate method.